evutil.c

```diff
@@ -194,11 +201,177 @@ evutil_read_file_(const char *filename, char **content_out, size_t *len_out,
+	if (strlcpy(listen_addr.sun_path, tmp_file, UNIX_PATH_MAX) >=
+		UNIX_PATH_MAX) {
+		event_warnx("Temp file name is too long");
+		goto tidy_up_and_fail;
+	}
+
+	if (bind(listener, (struct sockaddr *) &listen_addr, sizeof (listen_addr))
+		== -1)
+		goto tidy_up_and_fail;
+	if (listen(listener, 1) == -1)
+		goto tidy_up_and_fail;
+
+	connector = socket(family, type, 0);
+	if (connector < 0)
+		goto tidy_up_and_fail;
+
+	memset(&connect_addr, 0, sizeof(connect_addr));
+
+	/* We want to find out the port number to connect to.  */
+	size = sizeof(connect_addr);
+	if (getsockname(listener, (struct sockaddr *) &connect_addr, &size) == -1)
+		goto tidy_up_and_fail;
+
+	if (connect(connector, (struct sockaddr *) &connect_addr,
+				sizeof(connect_addr)) == -1)
+		goto tidy_up_and_fail;
+
+	size = sizeof(listen_addr);
+	acceptor = accept(listener, (struct sockaddr *) &listen_addr, &size);
+	if (acceptor < 0)
+		goto tidy_up_and_fail;
+	if (size != sizeof(listen_addr))
+		goto abort_tidy_up_and_fail;
+	/* Now check we are talking to ourself by matching port and host on the
+	   two sockets.	 */
+	if (getsockname(connector, (struct sockaddr *) &connect_addr, &size) == -1)
+		goto tidy_up_and_fail;
+
+	if (size != sizeof(connect_addr) ||
+	    listen_addr.sun_family != connect_addr.sun_family ||
+	    evutil_ascii_strcasecmp(listen_addr.sun_path, connect_addr.sun_path))
+		goto abort_tidy_up_and_fail;
+
+	evutil_closesocket(listener);
+	fd[0] = connector;
+	fd[1] = acceptor;
+
+	return 0;
+
+ abort_tidy_up_and_fail:
+	saved_errno = ERR(ECONNABORTED);
+ tidy_up_and_fail:
+	if (saved_errno < 0)
+		saved_errno = EVUTIL_SOCKET_ERROR();
+	if (listener != -1)
+		evutil_closesocket(listener);
+	if (connector != -1)
+		evutil_closesocket(connector);
+	if (acceptor != -1)
+		evutil_closesocket(acceptor);
+
+	DeleteFileA(tmp_file);
+	EVUTIL_SET_SOCKET_ERROR(saved_errno);
+	return -1;
+#undef ERR
+}
+
+static int
+evutil_win_socketpair(int family, int type, int protocol,
+    evutil_socket_t fd[2])
+{
+	/* The family only support AF_UNIX and AF_INET */
+	if (protocol || (family != AF_UNIX && family != AF_INET)) {
+		EVUTIL_SET_SOCKET_ERROR(WSAEAFNOSUPPORT);
+		return -1;
+	}
+	if (family == AF_UNIX && evutil_check_working_afunix_()) {
+	/* Win10 does not support AF_UNIX socket of a type other than
+	 * SOCK_STREAM still now. */
+		type = SOCK_STREAM;
+	} else {
+	/* If the family is set to AF_UNIX and it does not work, we will 
+	 * change it to AF_UNIX forcely. */
+		family = AF_INET;
+	}
+	if (family == AF_UNIX) {
+		return evutil_win_socketpair_afunix(family, type, protocol, fd);
+	} else {
+		return evutil_ersatz_socketpair_(family, type, protocol, fd);
+	}
+}
+#endif
+
 int
 evutil_socketpair(int family, int type, int protocol, evutil_socket_t fd[2])
 {
-#ifndef _WIN32
+#if !defined(_WIN32)
 	return socketpair(family, type, protocol, fd);
+#elif defined(EVENT__HAVE_AFUNIX_H)
+	return evutil_win_socketpair(family, type, protocol, fd);
```

Comment:
I think it is better to move checking for AF_UNIX and it's support here, to avoid extra branching